### PR TITLE
[Loupe Finding] Taker Taproot contract validation

### LINKED ISSUE #885: [Loupe Finding] bug: Taker does not fully validate maker Taproot contract data before accepting incoming swapcoin
## Description
The taker does not fully bind maker-provided Taproot contract data to the actual contract transaction it receives before accepting the maker's incoming swapcoin.

A malicious maker can send internally inconsistent `TaprootContractData`, for eg:
- claim a higher amount in `amounts[i]` than the actual `contract_txs[i].output[0].value`
- provide Taproot script data that is only partially shape-checked
- provide a `tap_tweak` value that is stored later but not verified against the internal key and script tree

## Steps to Reproduce
1. Run a Taproot swap with a modified/malicious maker.
2. Have the maker create a valid Taproot contract transaction with a lower output value than expected, for example `10_000 sats`.
3. Have the maker still advertise the normal post-fee amount in `TaprootContractData.amounts`.
4. Let the taker continue the swap flow.
5. Observe that the taker accepts the maker contract data, later fails to sweep the incoming swapcoin, and still reports the swap as successful.

## Expected Behavior
The taker should reject inconsistent maker Taproot contract data before accepting or storing an incoming swapcoin.

Specifically, it should verify that:

- `contract_txs.len() == amounts.len()`
- each `amounts[i]` equals `contract_txs[i].output[0].value`
- each contract transaction output pays to the Taproot scriptPubKey derived from the advertised internal key and script tree
- the hashlock script matches the exact expected template:
  `OP_SHA256 <hash> OP_EQUALVERIFY <pubkey> OP_CHECKSIG`
- the timelock script matches the exact expected template:
  `<locktime> OP_CLTV OP_DROP <pubkey> OP_CHECKSIG`
- the advertised `tap_tweak` matches the recomputed tweak from the internal key and script tree

## Actual Behavior
The taker can accept maker Taproot contract data that is not fully bound to the actual contract transaction output.

In a swap performed on signet, the maker contract output was only `0.0001 BTC`, but the taker later attempted to sweep `0.0029755 BTC(swap_amount = 30k sats)` from it:

```text
Failed to broadcast sweep tx for swapcoin 4b996d5c02a202125486432089728c319758a6b02715b26b2e6943994c131afd:
RpcError { code: -26, message: "bad-txns-in-belowout, value in (0.0001) < value out (0.0029755)" }

```
The swap report then printed SUCCESS even though no swap output was received:
```
TAKER SWAP REPORT

Swap ID           : db0f63af8be80091
Status            : SUCCESS
Target Amount     : 300000 sats
Total Output      : 0 sats
Total Fee Paid    : 300308 sats
Fee Percentage    : 100.1027%
```
Final taker balance, it lost `~30k` sats
```
{
  "regular": 9679384,
  "contract": 0,
  "swap": 0,
  "spendable": 9679384
}
```

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [x] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [x] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### PR DESCRIPTION
**Pull request by 0xEgao**
# Pull Request

## Description
Fixes #885, #906

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [x] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [ ] Neither (infrastructure / tooling only)

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [x] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [x] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [x] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [x] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [x] Integration tests pass (`cargo test --features integration-test`)
- [x] Changes were manually tested on **regtest**
- [x] End-to-end maker ↔ taker swap tested (where applicable)
- [x] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [x] Complex logic is commented

### Security & Privacy (Critical)
- [ ] This change preserves **trustlessness** and **atomic swap guarantees**
- [ ] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [ ] No new attack vectors or trust assumptions introduced
- [ ] Edge cases and error handling considered
- [ ] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [ ] `integration-test` feature flag is not reachable in production code paths

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved validation of Taproot swap setup to catch mismatched or underfunded contracts earlier.
  * Strengthened confirmation checks so transaction and fidelity verification use the actual block height instead of an estimated value.
  * Added clearer rejection of invalid proof data when confirmation details do not match on-chain results.
  * Incoming swap amounts are now derived more reliably from transaction outputs, reducing acceptance of incorrect values.

### CODE REVIEW

**Comment by pawurb** [comment]
Hi, thanks for using the hotpath CI integration! I'm starting to work on a new versions that produces significantly less noise and delivers actionable reviews for each PR. I'd love to get your feedback once it launches. I've added a waitlist on https://hotpath.rs/ if you're interested.

File: tests/integration/taproot_contract_validation.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    let summary = taker
+        .prepare_coinswap(swap_params)
+        .expect("failed to prepare Taproot coinswap");
+
+    // The taker must reject during maker contract verification, before storing
+    // an incoming swapcoin from the malicious contract data.
+    assert!(
+        taker.start_coinswap(&summary.swap_id).is_err(),
+        "taker must reject maker contract data that claims more than the tx output"
+    );
```
Comment (mojoX911) [inline tests/integration/taproot_contract_validation.rs:72]: We can assert for the exact error here?

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Ack. Very clean. Just one comment.

File: tests/integration/taproot_contract_validation.rs (hunk 1 above)
Comment (0xEgao) [reply tests/integration/taproot_contract_validation.rs:72]: done, now it assert the exact error.

**Review by mojoX911 — APPROVED** [review]
Ack

**Review by stark-3k — APPROVED** [review]
Ack, much needed checks
